File: python/kohaku/src/kohaku/spec/cache_key.py
"""Cache key and Spec hash (port of TS cache-key.ts).

The TS implementation is async due to WebCrypto, but Python's hashlib is synchronous, so these are
synchronous functions.
"""

from __future__ import annotations

import weakref
from dataclasses import dataclass

from .canonical_json import _utf16_key, canonical_stringify, sha256_hex
from .models import SPEC_VERSION, UISpec


@dataclass(frozen=True)
class CacheKeyParts:
    intentHash: str
    dataVersion: str
    catalogFingerprint: str | None = None
    """Component that prevents mixing up an old Spec when a component is revised."""
    specVersion: str | None = None
    generatorVersion: str | None = None
    """Generator version. An optional component that separates the cache on prompt revisions / model changes.

    If unspecified, matches the previous 5-component key exactly (prevents the cache from being invalidated on introduction).
    """
    policyFingerprint: str | None = None
    """A fingerprint of the ComposePolicy fields that affect prompt content but were not otherwise
    reflected in the cache key (outputLanguage / designSystem / fewShot.id / selectComponents.id — see
    composer.context.policy_fingerprint). Appended as the 7th component, after generatorVersion. If
    omitted (or the empty string), the key is completely unchanged from before this field existed. Because
    this component sits after generatorVersion positionally, a "-" placeholder is inserted in the
    generatorVersion slot when this is given but generatorVersion is not, so the two never collide on the
    same string for different inputs (mirrors TS spec-core's cacheKey exactly).
    """


def cache_key(parts: CacheKeyParts) -> str:
    """Spec cache key (Intent + data version, + catalog fingerprint)."""
    segments = [
        "kohaku",
        parts.specVersion if parts.specVersion is not None else SPEC_VERSION,
        parts.intentHash,
        parts.dataVersion,
        parts.catalogFingerprint if parts.catalogFingerprint is not None else "-",
    ]
    policy_fingerprint = parts.policyFingerprint
    has_policy_fingerprint = policy_fingerprint is not None and policy_fingerprint != ""
    # generatorVersion is appended as a 6th component when specified, or as a "-" placeholder when
    # policyFingerprint (the 7th component) is given without it — keeping the two positionally distinct so
    # a generatorVersion-only key can never collide with a policyFingerprint-only key. If neither is given,
    # the string matches the legacy 5-component key exactly.
    if parts.generatorVersion is not None or has_policy_fingerprint:
        segments.append(parts.generatorVersion if parts.generatorVersion is not None else "-")
    if has_policy_fingerprint and policy_fingerprint is not None:
        segments.append(policy_fingerprint)
    return ":".join(segments)


def combine_data_versions(entries: list[tuple[str, str]]) -> str:
    """Deterministically combine the dataVersions of multiple QueryHandles.

    entries are (uri, version) pairs. Fold identical versions: when multiple $refs point at the same
    data source, they normalize to a single version and can be reconciled directly against each $ref's
    response dataVersion. Without dedup, even identical versions would become `multi:`, and the
    renderer's dataVersion reconciliation would always judge STALE. This single-version shortcut is why
    the bare version string (never a `multi:` wrapper) is still returned whenever there is only one
    distinct version, even though the input is now uri-aware.

    The combined hash is uri-aware (`uri=version` pairs, sorted, then hashed) rather than a hash of the
    version values alone, so that {a: v1, b: v2} and {a: v2, b: v1} never collapse onto the same cache
    key merely because the same two version strings happen to appear.
    """
    unique = list(dict.fromkeys(version for _uri, version in entries))
    if len(unique) == 0:
        return "none"
    if len(unique) == 1:
        return unique[0]
    pairs = [f"{uri}={version}" for uri, version in entries]
    sorted_pairs = sorted(pairs, key=_utf16_key)
    hex_digest = sha256_hex("|".join(sorted_pairs))
    # 64 bits (16 hex) is enough: a multi: value never matches a single dataVersion in STALE judgment, so
    # collisions are harmless. Its birthday-collision room as a cache-key component has negligible impact.
    return f"multi:{hex_digest[:16]}"


def compute_spec_hash(spec: UISpec) -> str:
    """Deterministic hash of the whole Spec (used for View Lineage and the specHash reference in /events)."""
    return f"sha256:{sha256_hex(canonical_stringify(spec.to_wire()))}"


@dataclass
class _StructureHashMemoEntry:
    """One compute_structure_hash memo entry: the identity of the components/events/state it was computed
    against (a cheap staleness check on hit) plus the resulting hash."""

    components_id: int
    events_id: int
    state_id: int
    value: str


# Memoizes compute_structure_hash. A cache hit that returns the same in-memory UISpec object on repeated
# lookups (a common StoragePort.get_spec_cache shape) means re-hashing it on every single call (once per
# view.composed recording) is pure waste; this lets such repeats reuse the first computation.
#
# TS's structureHashMemo (cache-key.ts) is a `WeakMap<ComponentNode[], ...>` keyed on the Spec's `components`
# array reference. That exact shape does not carry over to Python: a plain `list` cannot be weakly referenced
# here (`weakref.ref([])` raises TypeError), and UISpec itself -- the only object in this pipeline pydantic
# actually allows a weak reference to -- has no `__hash__` (pydantic v2's BaseModel defines value-based
# `__eq__` without `__hash__`, so plain instances are unhashable and cannot be used as a
# `weakref.WeakKeyDictionary` key). Making UISpec hashable (`model_config = ConfigDict(frozen=True)`) was
# rejected as a fix: pydantic's frozen hash is computed from field *values*, which would force hashing the
# whole Spec structure on every lookup -- defeating the O(1) identity-check this memo exists for.
#
# So this is a plain dict keyed by `id(spec)` (the UISpec object's own identity) instead of a
# WeakKeyDictionary, with cleanup done manually via `weakref.finalize(spec, ...)` when the memo is written
# (UISpec itself IS weakly-referenceable, confirmed above) -- the same "no leak, no stale hit from an id()
# address reused by an unrelated later object" guarantee a WeakKeyDictionary would give, without requiring
# UISpec to become hashable. `components_id` / `events_id` / `state_id` are the secondary identity check on
# hit (mirroring TS's `cached.events === spec.events && cached.state === spec.state`): keying primarily on
# `id(spec)` rather than `id(spec.components)` narrows the hit rate slightly versus TS (two distinct UISpec
# wrapper objects sharing the same underlying components/events/state -- e.g. from `dataclasses.replace`-style
# rewrapping -- will not share a cache entry here), but that is a missed optimization, never a correctness
# risk: any actual reference reuse across genuinely different content is still caught by the identity check.
_structure_hash_memo: dict[int, _StructureHashMemoEntry] = {}


def compute_structure_hash(spec: UISpec) -> str:
    """Structure-only hash (components + events; provenance / dataVersion excluded).

    Used for the "structural stability" judgment of L1→L0 fixation. state is included in the input only when
    present (the hash of a Spec that does not use state stays completely unchanged from the previous value,
    protecting the fixation stability tally).

    Memoized by `spec`'s own identity (see `_structure_hash_memo`'s doc comment above for why this differs
    in shape, though not in intent, from TS's WeakMap-based memo).
    """
    key = id(spec)
    components_id, events_id, state_id = id(spec.components), id(spec.events), id(spec.state)
    cached = _structure_hash_memo.get(key)
    if (
        cached is not None
        and cached.components_id == components_id
        and cached.events_id == events_id
        and cached.state_id == state_id
    ):
        return cached.value
    value = _compute_structure_hash_uncached(spec)
    _structure_hash_memo[key] = _StructureHashMemoEntry(
        components_id=components_id, events_id=events_id, state_id=state_id, value=value
    )
    # Evict this entry once `spec` itself is garbage-collected, so the memo neither grows unboundedly nor
    # risks a later unrelated object reusing this exact `id()` and spuriously hitting a stale entry.
    weakref.finalize(spec, _structure_hash_memo.pop, key, None)
    return value


def _compute_structure_hash_uncached(spec: UISpec) -> str:
    structure: dict[str, object] = {
        "components": [c.to_wire() for c in spec.components],
        "events": [e.to_wire() for e in spec.events],
    }
    if spec.state is not None:
        structure["state"] = spec.state
    return f"sha256:{sha256_hex(canonical_stringify(structure))}"

File: packages/spec-core/src/cache-key.ts
import { canonicalStringify, sha256Hex } from "./canonical-json.js";
import type { ComponentNode } from "./schema/component.js";
import { SPEC_VERSION, type UISpec } from "./schema/spec.js";

export interface CacheKeyParts {
  intentHash: string;
  dataVersion: string;
  /** A component that keeps a stale Spec from being mistaken for a current one when the catalog is revised. */
  catalogFingerprint?: string;
  specVersion?: string;
  /**
   * The generator version. An optional component that separates the cache across prompt revisions or
   * model changes. If omitted, the key matches the legacy 5-component key exactly (so introducing this
   * component does not blow away every existing cache entry).
   */
  generatorVersion?: string;
  /**
   * A fingerprint of the ComposePolicy fields that affect prompt content but were not otherwise reflected
   * in the cache key (outputLanguage / designSystem / fewShot.id / selectComponents.id — see
   * @kohaku-ui/composer's policyFingerprint). Appended as the 7th component, after generatorVersion.
   * If omitted (or the empty string), the key is completely unchanged from before this field existed —
   * introducing it alone does not invalidate any existing cache entry. Because this component sits after
   * generatorVersion positionally, a "-" placeholder is inserted in the generatorVersion slot when this is
   * given but generatorVersion is not, so the two never collide on the same string for different inputs.
   */
  policyFingerprint?: string;
}

/** Spec cache key (Intent + data version, plus the catalog fingerprint). */
export function cacheKey(parts: CacheKeyParts): string {
  const segments = [
    "kohaku",
    parts.specVersion ?? SPEC_VERSION,
    parts.intentHash,
    parts.dataVersion,
    parts.catalogFingerprint ?? "-",
  ];
  const hasPolicyFingerprint = parts.policyFingerprint != null && parts.policyFingerprint !== "";
  // generatorVersion is appended as the 6th component when specified, or as a "-" placeholder when
  // policyFingerprint (the 7th component) is given without it — keeping the two positionally distinct so
  // a generatorVersion-only key can never collide with a policyFingerprint-only key. If neither is given,
  // the string matches the legacy 5-component key exactly, so introducing either component alone does not
  // invalidate every existing cache entry at once.
  if (parts.generatorVersion != null || hasPolicyFingerprint) {
    segments.push(parts.generatorVersion ?? "-");
  }
  if (hasPolicyFingerprint) segments.push(parts.policyFingerprint!);
  return segments.join(":");
}

/** One $ref's URI and its resolved dataVersion, the input unit for combineDataVersions. */
export interface RefVersionEntry {
  uri: string;
  version: string;
}

/**
 * Deterministically combines the dataVersion of multiple QueryHandles.
 * Collapses identical versions: when several $refs point at the same data source (a KPI list, etc.),
 * they normalize to a single version that can be matched directly against each $ref's response
 * dataVersion. Without deduplication, even identical versions would become a `multi:` value and the
 * Renderer's dataVersion matching would always judge STALE. This single-version shortcut is why the
 * function keeps returning the bare version string (never a `multi:` wrapper) whenever there is only
 * one distinct version, even though the input is now URI-aware.
 *
 * The combined hash is uri-aware (`uri=version` pairs, sorted, then hashed) rather than a hash of the
 * version values alone, so that `{a: v1, b: v2}` and `{a: v2, b: v1}` never collapse onto the same
 * cache key merely because the same two version strings happen to appear.
 */
export async function combineDataVersions(entries: RefVersionEntry[]): Promise<string> {
  const unique = [...new Set(entries.map((e) => e.version))];
  if (unique.length === 0) return "none";
  if (unique.length === 1) return unique[0]!;
  const pairs = entries.map((e) => `${e.uri}=${e.version}`);
  const sorted = [...pairs].sort();
  const hex = await sha256Hex(sorted.join("|"));
  // 64 bits (16 hex) is enough: a multi: value represents several distinct data versions and, in
  // STALE checks, never matches a single dataVersion, so a collision is harmless. As a cache-key
  // component a theoretical birthday-collision margin remains, but the impact is negligible.
  return `multi:${hex.slice(0, 16)}`;
}

/** Deterministic hash of the whole Spec (used for View Lineage and the specHash reference on /events). */
export async function computeSpecHash(spec: UISpec): Promise<string> {
  const hex = await sha256Hex(canonicalStringify(spec));
  return `sha256:${hex}`;
}

/** One computeStructureHash memo entry: the events/state it was computed against (for a cheap staleness check) + the resulting promise. */
interface StructureHashMemo {
  events: UISpec["events"];
  state: UISpec["state"];
  promise: Promise<string>;
}

/**
 * Memoizes computeStructureHash by the Spec's `components` array reference. A cache hit that returns the
 * same in-memory UISpec object on repeated lookups (a common StoragePort.getSpecCache shape) keeps that
 * object's `components` array reference stable across requests, so re-hashing it on every single request
 * (once per view.composed recording) is pure waste — this lets such repeats reuse the first computation.
 * A Spec whose storage round-trips through JSON (a fresh array each time) simply never hits the cache here
 * and pays the same cost as before it existed (no correctness risk either way).
 *
 * Keyed only by the `components` reference, with `events`/`state` reference-checked on hit: if either
 * differs from what the cached promise was computed against (a components array reused across genuinely
 * different Specs), the entry is recomputed rather than trusted — so a rare reference reuse can never
 * silently return a stale hash for the wrong content.
 */
const structureHashMemo = new WeakMap<ComponentNode[], StructureHashMemo>();

/**
 * Structure-only hash (components + events; excludes provenance / dataVersion).
 * Used to judge the "structural stability" of L1→L0 fixation: how consistently the same Intent
 * emits the same structure across different dataVersions is measured by this hash's match rate.
 *
 * state is included in the input **only when it exists**. The hash of an existing Spec without state
 * (0.1-era, or 0.2 that does not use state) stays completely unchanged from the legacy value,
 * protecting the fixation stability tally. Since state is the very branch point of the display
 * condition (visibleWhen), when declared it is folded in as part of the structure.
 */
export function computeStructureHash(spec: UISpec): Promise<string> {
  const cached = structureHashMemo.get(spec.components);
  if (cached != null && cached.events === spec.events && cached.state === spec.state) {
    return cached.promise;
  }
  const promise = computeStructureHashUncached(spec);
  structureHashMemo.set(spec.components, { events: spec.events, state: spec.state, promise });
  return promise;
}

/** The uncached computation body, kept `async` so a synchronous canonicalStringify throw (a non-finite
 * number, etc.) surfaces as a rejected Promise exactly as it did before memoization was added, rather than
 * escaping computeStructureHash as a synchronous throw. */
async function computeStructureHashUncached(spec: UISpec): Promise<string> {
  const structure =
    spec.state != null
      ? { components: spec.components, events: spec.events, state: spec.state }
      : { components: spec.components, events: spec.events };
  const hex = await sha256Hex(canonicalStringify(structure));
  return `sha256:${hex}`;
}
